perltalk.tex

```diff
@@ -743,7 +743,7 @@ \subsubsection{Writing generic mappers with \texttt{Moose} and roles}
     my $orig = shift;
     my $self = shift;
     return sub {
-        return $self->$orig->( $self, @_ );
+        return $self->$orig( @_ );
```

Comment:
I **think** this is right - could you try it to be sure?

Reply:
Aha, you ran into my trick to pass $self for singe file module development. The text as originally written is correct, but the context in the preprint may be too short to interpret. 
Consider a module X with a method Y. The Y method would parse its arguments as  : 
```perl
Y sub { 
    $self = shift
   ## all the arguments that Y needs can be parsed now
...
}
```
Ordinarily, one would have a file X.pm around and another file testdrive.pl that one would use to test their implementation, tweak the code etc. Every time one finds out that Y does something wrong, one would have to rebuild the module X.pm and switch back and forth between the X.pm and testdrive.pl files. 

What I did, is to create an intermediate package Xalt, which accepts coderefs as arguments. Then inside the testdrive.pl one can do this:
```perl
use Moose;
my $obj = Xalt->new(
Y = \&Y_candidate
);

Y_candidate sub {
    $self = shift
   ## all the arguments that Y needs can be parsed now
...
}
```
The code you highlighted is found in the implementation of the Xalt module and it intends to make all candidate methods $self aware (as they would have been if they were found in the X.pm file) . This trick allows one to develop all methods for X, inside the testdrive.pl using the Xalt.pm module. Once the code for these functions have been developed, one simply copies the code verbatim to the X.pm and they are set. The Xalt.pm module can be discarded, or kept around for future development for X.pm. 

Examples of how this was trick was used may be found in the repo : [Enhancing edlib](https://github.com/chrisarg/bio-seqalignment-examples-enhancing-edlib) ; inside the scripts directory [testMapMCE_overC.pl](https://github.com/chrisarg/bio-seqalignment-examples-enhancing-edlib/blob/main/scripts/testMapMCE_overC.pl) is an example of a testdrive.pl file. 
The corresponding "X.pm" is the [Generic.pm](https://github.com/chrisarg/bio-seqalignment-components-seqmapping/blob/main/lib/Bio/SeqAlignment/Components/SeqMapping/Mapper/Generic.pm) in this [repo](https://github.com/chrisarg/bio-seqalignment-components-seqmapping/).

Reply:
That's one way to approach such a situation. I assume you actually mean:

```perl
sub Y {
  $self = shift
...
}
...
my $obj = Xalt->new(
  Y => \&Y_candidate
);
```

Rather than have to do rebuild on a module that's not doing what I need (I assume you mean build and also install), sometimes I do:

```
chmod u+w `perldoc -l Module::Name`
vi `perldoc -l Module::Name`
```

and then keep testing the consuming module until it's right, then put those changes in the changed module's source. Another, more idiomatic way is just to add tests to the module that needs changing so its required behaviour (aka "contract") is correct for the consuming module. I have to do that sort of thing all the time when something obscure in PDL breaks a downstream module. Your way clearly works for you, and that's great, but there are other ways.